vault/resource_raft_snapshot_agent_config.go

```diff
@@ -0,0 +1,498 @@
+		"storage_type": d.Get("storage_type"),
+	}
+
+	if storageType == "local" {
+		if v, ok := d.GetOk("local_max_space"); ok && v != 0 {
+			data["local_max_space"] = v
+		} else {
+			return nil, errors.New("specified local storage without setting local_max_space")
+		}
+
+	}
+
+	if storageType == "aws-s3" {
+		if v, ok := d.GetOk("aws_s3_bucket"); ok && v != "" {
+			data["aws_s3_bucket"] = v
+		} else {
+			return nil, errors.New("specified aws-s3 storage without setting aws_s3_bucket")
+		}
+		if v, ok := d.GetOk("aws_s3_region"); ok && v != "" {
+			data["aws_s3_region"] = v
+		} else {
+			return nil, errors.New("specified aws-s3 storage without setting aws_s3_region")
+		}
+		if v, ok := d.GetOk("aws_access_key_id"); ok {
+			data["aws_access_key_id"] = v
+		}
+		if v, ok := d.GetOk("aws_session_token"); ok {
+			data["aws_session_token"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_endpoint"); ok {
+			data["aws_s3_endpoint"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_disable_tls"); ok {
+			data["aws_s3_disable_tls"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_force_path_style"); ok {
+			data["aws_s3_force_path_style"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_enable_kms"); ok {
+			data["aws_s3_enable_kms"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_server_side_encryption"); ok {
+			data["aws_s3_server_side_encryption"] = v
+		}
+		if v, ok := d.GetOk("aws_s3_kms_key"); ok {
+			data["aws_s3_kms_key"] = v
+		}
+	}
+
+	if storageType == "google-gcs" {
+		if v, ok := d.GetOk("google_gcs_bucket"); ok && v != "" {
+			data["google_gcs_bucket"] = v
+		} else {
+			return nil, errors.New("specified google-gcs storage without setting google_gcs_bucket")
+		}
+		if v, ok := d.GetOk("google_service_account_key"); ok {
+			data["google_service_account_key"] = v
+		}
+		if v, ok := d.GetOk("google_endpoint"); ok {
+			data["google_endpoint"] = v
+		}
+		if v, ok := d.GetOk("google_disable_tls"); ok {
+			data["google_disable_tls"] = v
+		}
+	}
+
+	if storageType == "azure-blob" {
+		if v, ok := d.GetOk("azure_container_name"); ok && v != "" {
+			data["azure_container_name"] = v
+		} else {
+			return nil, errors.New("specified azure-blob storage without setting azure_container_name")
+		}
+		if v, ok := d.GetOk("azure_account_name"); ok {
+			data["azure_account_name"] = v
+		}
+		if v, ok := d.GetOk("azure_account_key"); ok {
+			data["azure_account_key"] = v
+		}
+		if v, ok := d.GetOk("azure_blob_environment"); ok {
+			data["azure_blob_environment"] = v
+		}
+		if v, ok := d.GetOk("azure_endpoint"); ok {
+			data["azure_endpoint"] = v
+		}
+	}
+	return data, nil
+}
+
+func createOrUpdateSnapshotAgentConfigResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Get("name").(string)
+	path := fmt.Sprintf(snapshotAutoPath, name)
+
+	config, err := buildConfigFromResourceData(d)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Configuring automatic snapshots: %q", name)
```

Comment:
I don't think we can use `Logical().Write()` here because under the hood it is making a `PUT` request ([code](https://github.com/hashicorp/vault/blob/main/api/logical.go#L133)) and the Snapshot Agent API is expecting a `POST` ([docs](https://www.vaultproject.io/api-docs/system/storage/raftautosnapshots#create-update-an-automated-snapshots-config)).

Reply:
Most Vault endpoints (including this one) support both `PUT` and `POST`. I tested the above code and it works, I think it's better to change to a `Logical().Write()` to maintain consistency with other resources in the provider

Reply:
That makes my inner API architecture senses _very_ sad, but if it works... I guess it works. 😜 Swapped.